packages/examples/src/diffie-hellman/diffie-hellman.ts

```diff
@@ -0,0 +1,158 @@
+// Ported from the HasChor implementation here: https://github.com/gshen42/HasChor/blob/42ae1ef9a500dadd82f0cfe5dee3c2aa631d8f4d/examples/diffiehellman/Main.hs
+
+//////////////////////
+// Helper functions //
+/////////////////////
+
+// Return all divisors of `x`
+const divisors = (x: number): number[] => {
+  let divs: number[] = [1, x];
+  for (let y = 2; y <= x / 2; y++) {
+    if (x % y == 0) divs.push(y);
+  }
+  return divs;
+};
+
+// Checks if input number is prime
+const isPrime = (x: number): boolean => {
+  let divs: number[] = divisors(x);
+  if (divs.length == 2) return true;
+  return false;
+};
+
+// Prime number generator
+const primeNums = function* () {
+  let y = 1;
+  while (y++ > 0) {
+    if (isPrime(y)) yield y;
+  }
+};
+
+// Return random integer in range [min, max]
+// https://stackoverflow.com/a/7228322
+const randInRange = (min: number, max: number) =>
+  Math.floor(Math.random() * (max - min + 1) + min);
+
+///////////////////////////////////
+// Diffie-Hellman implementation //
+///////////////////////////////////
+const locations = ["alice", "bob"];
+export type Location = (typeof locations)[number];
+
+export const diffieHellman = <A extends Location, B extends Location>(
+  a: A,
+  b: B
+) => {
+  const choreo: Choreography<
+    Location,
+    [Located<boolean, A>], // Specify boolean argument specifically for location "A" ("A|B" instead would mean for both)
+    [Located<number, A>, Located<number, B>]
+  > = async ({ locally, comm }, [arg]) => {
+    // The `arg` specifies whether `a` should wait to initiate key exchange
+
+    // Wait for alice to start key exchange (if desired)
+    await locally(a, (unwrap: any) => {
+      console.log("Press enter to begin key exchange...");
+      // Check for key input on stdin in node: https://stackoverflow.com/a/72906729
+      let status: boolean = !unwrap(arg);
+      if (!status) {
+        // If waiting is desired
+        process.stdin
+          .setRawMode(true)
+          .setEncoding("utf8")
+          .resume()
+          .on("data", (key: string) => {
+            if (key.charCodeAt(0) == 0xd) {
+              process.stdin.setRawMode(false).pause();
+              status = true;
+            } else if (key.charCodeAt(0) == 0x3) {
+              process.exit(); // Ctrl-c
+            }
+          });
+      }
+      return new Promise<void>((resolve) => {
+        let id = setInterval(() => {
+          if (status) {
+            clearInterval(id); // Clear the interval checking
+            resolve(); // Begin key exchange
+          }
+        });
+      });
+    });
+    await locally(b, () =>
+      console.log("Waiting for alice to begin exchange...")
+    );
+
+    // Alice picks p and g to send to bob
+    const pa = await locally(a, () => {
+      const x = randInRange(200, 1000); // Random number in range [200, 1000]
+      const primeGen = primeNums();
+      let prime = primeGen.next().value;
```

Comment:
not a big deal, but `prime` is typed `number | void`. We want this to be just `number`? Looks like it has to do with TypeScript and generators.

https://stackoverflow.com/questions/70545038/how-to-avoid-void-type-in-generators-typescript

Reply:
Yeah, it seems like TS thinks the generator might eventually run out of values when it clearly won't. It shouldn't be an issue though since the generator will clearly never return nothing

Reply:
I'm not worried about the generator running out of values but `void` makes it not type check. If that is intentional, you should add explicit cast (`as number`).

Reply:
Ok, I just made the fix then. Type assertions are neat...